[Inductor] Prune configs that require more shared memory than the hardware limit

wychi opened the pull request:
Summary:
This diff removes configs that require more shared memory than the hardware limit, which causes the following compilation error:
```
No valid triton configs. OutOfMemoryError: out of resource: triton_mm Required: 327680 Hardware limit:232448 Reducing block sizes or `num_stages` may help.
```

Test Plan:
```
buck2 test mode/dev-nosan fbcode//caffe2/test/inductor:max_autotune -- test_max_autotune_prune_choices -v 1,stderr
```

Rollback Plan:

Differential Revision: D80594562




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

wychi commented:
@pytorchbot label "topic: not user facing"

eellison commented on test/inductor/test_max_autotune.py:
```diff
@@ -1680,6 +1680,26 @@ def mm(x, y):
+        compiled_f(x, y)
+
+        self.assertEqual(
+            counters["inductor"]["select_algorithm_num_precompilation_exceptions"], 0
```
add another test where we skip a config based on shared mem?

wychi replied:
good idea. 

I'll add another test to `test_triton_heuristics.py` to test the configs directly.

eellison reviewed (approved):
looks good to me

wychi committed:
[Inductor] Prune configs that require more shared memory than the hardware limit (#161040)

Summary:

This diff removes configurations that exceed the hardware shared memory limit, which causes the following compilation error:
```
No valid triton configs. OutOfMemoryError: out of resource: triton_mm Required: 327680 Hardware limit:232448 Reducing block sizes or `num_stages` may help.
```

Test Plan:
```
buck2 test mode/dev-nosan fbcode//caffe2/test/inductor:max_autotune -- test_max_autotune_prune_choices -v 1,stderr
```

Rollback Plan:

Reviewed By: eellison

Differential Revision: D80594562

coconutruben commented:
@pytorchbot rebase

coconutruben commented:
@pytorchbot merge

coconutruben commented:
pretty sure the failure that prevented this earlier was a bug last week that's been fixed - rebasing and merging now again as this is already landed internally, so causing divergence when developing on top of template_heuristics

jeffdaily commented:
@pytorchbot revert -c nosignal -m "still failing on rocm, see https://hud.pytorch.org/failure?name=rocm%20%2F%20linux-jammy-rocm-py3.10%20%2F%20test%20(default%2C%203%2C%206%2C%20linux.rocm.gpu.2)&jobName=undefined&failureCaptures=inductor%2Ftest_triton_heuristics.py%3A%3ATestTritonHeuristics%3A%3Atest_prune_configs_over_shared_memory_limit_do_pruning_True"
